File: src/test/java/com/example/the_machine/service/RunServiceTest.java
package com.example.the_machine.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.example.the_machine.domain.EventType;
import com.example.the_machine.domain.RunEntity;
import com.example.the_machine.domain.RunFactory;
import com.example.the_machine.domain.ThreadEntity;
import com.example.the_machine.dto.AuthorType;
import com.example.the_machine.dto.CreateRunRequestDto;
import com.example.the_machine.dto.EventEnvelopeDto;
import com.example.the_machine.dto.MessageCreateRequestDto;
import com.example.the_machine.dto.RunDto;
import com.example.the_machine.dto.RunKind;
import com.example.the_machine.dto.RunStatus;
import com.example.the_machine.repo.RunRepository;
import com.example.the_machine.repo.ThreadRepository;
import com.example.the_machine.service.events.EventPublisher;
import com.example.the_machine.service.mapping.RunMapper;
import com.example.the_machine.service.mapping.ThreadMapper;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.context.ApplicationEventPublisher;

@ExtendWith(MockitoExtension.class)
class RunServiceTest {

  @Mock
  private RunRepository runRepository;

  @Mock
  private ThreadRepository threadRepository;

  @Mock
  private RunMapper runMapper;

  @Mock
  private ThreadMapper threadMapper;

  @Mock
  private EventPublisher eventPublisher;

  @Mock
  private ObjectMapper objectMapper;

  @Mock
  private ExecutorService runExecutorService;

  @Mock
  private AssistantEngine assistantEngine;

  @Mock
  private RunFactory runFactory;

  @Mock
  private ApplicationEventPublisher applicationEventPublisher;

  @InjectMocks
  private RunService runService;

  private UUID threadId;
  private UUID assistantId;
  private ThreadEntity thread;
  private RunEntity savedRun;
  private RunDto runDTO;
  private JsonNode mockJsonNode;

  @BeforeEach
  void setUp() {
    threadId = UUID.randomUUID();
    assistantId = UUID.randomUUID();

    thread = ThreadEntity.builder()
        .id(threadId)
        .title("Test Thread")
        .createdAt(Instant.now())
        .lastActivityAt(Instant.now())
        .build();

    savedRun = new RunEntity();
    savedRun.setId(UUID.randomUUID());
    savedRun.setThreadId(threadId);
    savedRun.setAssistantId(assistantId);
    savedRun.setStatus(RunEntity.Status.QUEUED);
    savedRun.setKind(RunEntity.Kind.FULL);

    runDTO = new RunDto(
        savedRun.getId(),
        savedRun.getThreadId(),
        savedRun.getAssistantId(),
        RunStatus.valueOf(savedRun.getStatus().name()),
        RunKind.valueOf(savedRun.getKind().name()),
        savedRun.getReason(),
        savedRun.getStartedAt(),
        savedRun.getCompletedAt(),
        null, // usage
        savedRun.getError()
    );

    mockJsonNode = new ObjectMapper().createObjectNode();
  }

  @Test
  void createRun_WithValidRequest_CreatesRunAndLogsEvent() {
    // Given
    MessageCreateRequestDto messageInput = new MessageCreateRequestDto(AuthorType.USER, null, null);
    CreateRunRequestDto request = new CreateRunRequestDto(threadId, assistantId, messageInput);

    when(threadRepository.findById(threadId)).thenReturn(Optional.of(thread));
    when(runFactory.createAndSaveRun(any(CreateRunRequestDto.class), eq(threadId))).thenReturn(
        savedRun.getId());
    when(runFactory.findManagedRun(savedRun.getId())).thenReturn(savedRun);
    when(runMapper.toDto(savedRun)).thenReturn(runDTO);
    when(objectMapper.valueToTree(runDTO)).thenReturn(mockJsonNode);

    // When
    RunDto result = runService.createRun(request);

    // Then
    assertThat(result).isEqualTo(runDTO);

    // Verify run entity was created through factory
    ArgumentCaptor<CreateRunRequestDto> requestCaptor = ArgumentCaptor.forClass(
        CreateRunRequestDto.class);
    verify(runFactory).createAndSaveRun(requestCaptor.capture(), eq(threadId));
    verify(runFactory).findManagedRun(savedRun.getId());

    CreateRunRequestDto capturedRequest = requestCaptor.getValue();
    assertThat(capturedRequest.threadId()).isEqualTo(threadId);
    assertThat(capturedRequest.assistantId()).isEqualTo(assistantId);

    // Verify event was logged
    ArgumentCaptor<EventEnvelopeDto> eventCaptor = ArgumentCaptor.forClass(EventEnvelopeDto.class);
    verify(eventPublisher).publishAndStore(eq(threadId), eventCaptor.capture(), eq(null));

    EventEnvelopeDto capturedEvent = eventCaptor.getValue();
    assertThat(capturedEvent.type()).isEqualTo(EventType.RUN_CREATED);
    assertThat(capturedEvent.threadId()).isEqualTo(threadId);
    assertThat(capturedEvent.runId()).isEqualTo(savedRun.getId());
    assertThat(capturedEvent.messageId()).isNull();

    // Verify event was published for async execution
    verify(applicationEventPublisher).publishEvent(
        any(com.example.the_machine.service.events.RunCreatedEvent.class));
  }

  @Test
  void createRun_WithoutInput_CreatesRunSuccessfully() {
    // Given
    CreateRunRequestDto request = new CreateRunRequestDto(threadId, assistantId, null);

    when(threadRepository.findById(threadId)).thenReturn(Optional.of(thread));
    when(runFactory.createAndSaveRun(any(CreateRunRequestDto.class), eq(threadId))).thenReturn(
        savedRun.getId());
    when(runFactory.findManagedRun(savedRun.getId())).thenReturn(savedRun);
    when(runMapper.toDto(savedRun)).thenReturn(runDTO);
    when(objectMapper.valueToTree(runDTO)).thenReturn(mockJsonNode);

    // When
    RunDto result = runService.createRun(request);

    // Then
    assertThat(result).isEqualTo(runDTO);
    verify(runFactory).createAndSaveRun(any(CreateRunRequestDto.class), eq(threadId));
    verify(eventPublisher).publishAndStore(eq(threadId), any(EventEnvelopeDto.class), eq(null));
    verify(applicationEventPublisher).publishEvent(
        any(com.example.the_machine.service.events.RunCreatedEvent.class));
  }

  @Test
  void createRun_WithNullAssistantId_UsesDefaultAssistant() {
    // Given
    CreateRunRequestDto request = new CreateRunRequestDto(threadId, null, null);

    when(threadRepository.findById(threadId)).thenReturn(Optional.of(thread));
    when(runFactory.createAndSaveRun(any(CreateRunRequestDto.class), eq(threadId))).thenReturn(
        savedRun.getId());
    when(runFactory.findManagedRun(savedRun.getId())).thenReturn(savedRun);
    when(runMapper.toDto(savedRun)).thenReturn(runDTO);
    when(objectMapper.valueToTree(runDTO)).thenReturn(mockJsonNode);

    // When
    runService.createRun(request);

    // Then
    ArgumentCaptor<CreateRunRequestDto> requestCaptor = ArgumentCaptor.forClass(
        CreateRunRequestDto.class);
    verify(runFactory).createAndSaveRun(requestCaptor.capture(), eq(threadId));

    CreateRunRequestDto capturedRequest = requestCaptor.getValue();
    assertThat(capturedRequest.assistantId()).isNull(); // Factory handles default assignment
  }

  @Test
  void createRun_WithInvalidThreadId_ThrowsException() {
    // Given
    UUID invalidThreadId = UUID.randomUUID();
    CreateRunRequestDto request = new CreateRunRequestDto(invalidThreadId, assistantId, null);

    when(threadRepository.findById(invalidThreadId)).thenReturn(Optional.empty());

    // When & Then
    assertThatThrownBy(() -> runService.createRun(request))
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessage("Thread not found: " + invalidThreadId);
  }

  @Test
  void createRun_SetsCorrectRunDefaults() {
    // Given
    CreateRunRequestDto request = new CreateRunRequestDto(threadId, assistantId, null);

    when(threadRepository.findById(threadId)).thenReturn(Optional.of(thread));
    when(runFactory.createAndSaveRun(any(CreateRunRequestDto.class), eq(threadId))).thenReturn(
        savedRun.getId());
    when(runFactory.findManagedRun(savedRun.getId())).thenReturn(savedRun);
    when(runMapper.toDto(savedRun)).thenReturn(runDTO);
    when(objectMapper.valueToTree(runDTO)).thenReturn(mockJsonNode);

    // When
    runService.createRun(request);

    // Then
    ArgumentCaptor<CreateRunRequestDto> requestCaptor = ArgumentCaptor.forClass(
        CreateRunRequestDto.class);
    verify(runFactory).createAndSaveRun(requestCaptor.capture(), eq(threadId));

    CreateRunRequestDto capturedRequest = requestCaptor.getValue();
    assertThat(capturedRequest.threadId()).isEqualTo(threadId);
    assertThat(capturedRequest.assistantId()).isEqualTo(assistantId);
    // Factory handles setting correct defaults (QUEUED, FULL, ID generation)
  }
}

File: src/test/kotlin/com/example/the_machine/service/RunServiceTest.kt
package com.example.the_machine.service

import com.example.the_machine.common.KotlinSerializationConfig
import com.example.the_machine.common.UuidV7
import com.example.the_machine.config.TestDataConfiguration.FIXED_INSTANT
import com.example.the_machine.config.TestDataConfiguration.FIXED_INSTANT_LATER
import com.example.the_machine.config.TestDataConfiguration.TEST_ASSISTANT_ID
import com.example.the_machine.config.TestDataConfiguration.TEST_INVALID_ID
import com.example.the_machine.config.TestDataConfiguration.TEST_RUN_ID
import com.example.the_machine.config.TestDataConfiguration.TEST_THREAD_ID
import com.example.the_machine.domain.BaseEntity
import com.example.the_machine.domain.EventType
import com.example.the_machine.domain.RunEntity
import com.example.the_machine.domain.ThreadEntity
import com.example.the_machine.dto.*
import com.example.the_machine.repo.RunRepository
import com.example.the_machine.repo.ThreadRepository
import com.example.the_machine.service.events.EventPublisher
import com.example.the_machine.service.events.RunCreatedEvent
import com.example.the_machine.service.mapping.RunMapper
import com.example.the_machine.service.mapping.ThreadMapper
import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.buildJsonObject
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.context.ApplicationEventPublisher
import java.util.*
import java.util.concurrent.ExecutorService

@ExtendWith(MockKExtension::class)
class RunServiceTest {

  @MockK
  @Suppress("unused") // Required for @InjectMockKs dependency injection
  private lateinit var runRepository: RunRepository

  @MockK
  private lateinit var threadRepository: ThreadRepository

  @MockK
  private lateinit var runMapper: RunMapper

  @MockK
  @Suppress("unused") // Required for @InjectMockKs dependency injection
  private lateinit var threadMapper: ThreadMapper

  @MockK(relaxed = true)
  private lateinit var eventPublisher: EventPublisher

  @MockK
  @Suppress("unused") // Required for @InjectMockKs dependency injection
  private lateinit var runExecutorService: ExecutorService

  @MockK
  @Suppress("unused") // Required for @InjectMockKs dependency injection
  private lateinit var assistantEngine: AssistantEngine

  @MockK(relaxed = true)
  private lateinit var applicationEventPublisher: ApplicationEventPublisher

  @MockK
  @Suppress("unused") // Required for @InjectMockKs dependency injection
  private lateinit var artifactService: ArtifactService

  @InjectMockKs
  private lateinit var runService: RunService

  private lateinit var threadId: UUID
  private lateinit var assistantId: UUID
  private lateinit var thread: ThreadEntity
  private lateinit var savedRun: RunEntity
  private lateinit var runDTO: RunDto
  private lateinit var mockJsonElement: JsonElement

  @BeforeEach
  fun setUp() {
    threadId = TEST_THREAD_ID
    assistantId = TEST_ASSISTANT_ID

    thread = ThreadEntity(
      title = "Test Thread",
      lastActivityAt = FIXED_INSTANT
    ).apply {
      // Use reflection to set the ID to our test constant
      val idField = BaseEntity::class.java.getDeclaredField("id")
      idField.isAccessible = true
      idField[this] = TEST_THREAD_ID
    }

    savedRun = RunEntity(
      threadId = threadId,
      assistantId = assistantId,
      status = RunEntity.Status.QUEUED,
      kind = RunEntity.Kind.FULL
    )

    runDTO = RunDto(
      savedRun.id,
      savedRun.threadId,
      savedRun.assistantId,
      RunStatus.valueOf(savedRun.status.name),
      RunKind.valueOf(savedRun.kind.name),
      savedRun.reason ?: "",
      savedRun.startedAt ?: FIXED_INSTANT,
      savedRun.completedAt ?: FIXED_INSTANT_LATER,
      buildJsonObject { },
      savedRun.error ?: ""
    )

    mockJsonElement = KotlinSerializationConfig.staticJson.parseToJsonElement("{}")
  }

  @Test
  fun createRun_WithValidRequest_CreatesRunAndLogsEvent() {
    mockkObject(UuidV7)
    try {
      // Given
      val messageInput = MessageCreateRequestDto(AuthorType.USER, emptyList(), emptyList())
      val request = CreateRunRequestDto(threadId, assistantId, messageInput)

      every { UuidV7.newId() } returns TEST_RUN_ID
      every { threadRepository.findById(threadId) } returns Optional.of(thread)
      every { runRepository.save(any<RunEntity>()) } returns savedRun
      every { runMapper.toDto(savedRun) } returns runDTO
      every { eventPublisher.publishAndStore(any(), any(), any()) } just Runs
      every { applicationEventPublisher.publishEvent(any()) } just Runs

      // When
      val result = runService.createRun(request)

      // Then
      assertThat(result).isEqualTo(runDTO)

      // Verify run entity was created and saved
      val runEntitySlot = slot<RunEntity>()
      verify { runRepository.save(capture(runEntitySlot)) }

      val capturedRunEntity = runEntitySlot.captured
      assertThat(capturedRunEntity.threadId).isEqualTo(threadId)
      assertThat(capturedRunEntity.assistantId).isEqualTo(assistantId)
      assertThat(capturedRunEntity.status).isEqualTo(RunEntity.Status.QUEUED)
      assertThat(capturedRunEntity.kind).isEqualTo(RunEntity.Kind.FULL)

      // Verify event was logged
      val eventSlot = slot<EventEnvelopeDto>()
      verify { eventPublisher.publishAndStore(threadId, capture(eventSlot), null) }

      val capturedEvent = eventSlot.captured
      assertThat(capturedEvent.type).isEqualTo(EventType.RUN_CREATED)
      assertThat(capturedEvent.threadId).isEqualTo(threadId)
      assertThat(capturedEvent.runId).isEqualTo(savedRun.id)
      assertThat(capturedEvent.messageId).isNull()

      // Verify event was published for async execution
      verify { applicationEventPublisher.publishEvent(any<RunCreatedEvent>()) }
    } finally {
      unmockkObject(UuidV7)
    }
  }

  @Test
  fun createRun_WithoutInput_CreatesRunSuccessfully() {
    // Given
    val request = CreateRunRequestDto(threadId, assistantId, null)

    every { threadRepository.findById(threadId) } returns Optional.of(thread)
    every { runRepository.save(any<RunEntity>()) } returns savedRun
    every { runMapper.toDto(savedRun) } returns runDTO
    every { eventPublisher.publishAndStore(any(), any(), any()) } just Runs
    every { applicationEventPublisher.publishEvent(any()) } just Runs

    // When
    val result = runService.createRun(request)

    // Then
    assertThat(result).isEqualTo(runDTO)
    verify { runRepository.save(any<RunEntity>()) }
    verify { eventPublisher.publishAndStore(any(), any(), any()) }
    verify { applicationEventPublisher.publishEvent(any<RunCreatedEvent>()) }
  }

  @Test
  fun createRun_WithNullAssistantId_UsesDefaultAssistant() {
    // Given
    val request = CreateRunRequestDto(threadId, null, null)

    every { threadRepository.findById(threadId) } returns Optional.of(thread)
    every { runRepository.save(any<RunEntity>()) } returns savedRun
    every { runMapper.toDto(savedRun) } returns runDTO
    every { eventPublisher.publishAndStore(any(), any(), any()) } just Runs
    every { applicationEventPublisher.publishEvent(any()) } just Runs

    // When
    runService.createRun(request)

    // Then
    val runEntitySlot = slot<RunEntity>()
    verify { runRepository.save(capture(runEntitySlot)) }

    val capturedRunEntity = runEntitySlot.captured
    assertThat(capturedRunEntity.assistantId).isNotNull() // Service handles default assignment
  }

  @Test
  fun createRun_WithInvalidThreadId_ThrowsException() {
    // Given
    val invalidThreadId = TEST_INVALID_ID
    val request = CreateRunRequestDto(invalidThreadId, assistantId, null)

    every { threadRepository.findById(invalidThreadId) } returns Optional.empty()

    // When & Then
    assertThatThrownBy { runService.createRun(request) }
      .isInstanceOf(IllegalArgumentException::class.java)
      .hasMessage("Thread not found: $invalidThreadId")
  }

  @Test
  fun createRun_SetsCorrectRunDefaults() {
    // Given
    val request = CreateRunRequestDto(threadId, assistantId, null)

    every { threadRepository.findById(threadId) } returns Optional.of(thread)
    every { runRepository.save(any<RunEntity>()) } returns savedRun
    every { runMapper.toDto(savedRun) } returns runDTO
    every { eventPublisher.publishAndStore(any(), any(), any()) } just Runs
    every { applicationEventPublisher.publishEvent(any()) } just Runs

    // When
    runService.createRun(request)

    // Then
    val runEntitySlot = slot<RunEntity>()
    verify { runRepository.save(capture(runEntitySlot)) }

    val capturedRunEntity = runEntitySlot.captured
    assertThat(capturedRunEntity.threadId).isEqualTo(threadId)
    assertThat(capturedRunEntity.assistantId).isEqualTo(assistantId)
    assertThat(capturedRunEntity.status).isEqualTo(RunEntity.Status.QUEUED)
    assertThat(capturedRunEntity.kind).isEqualTo(RunEntity.Kind.FULL)
    assertThat(capturedRunEntity.id).isNotNull()
    // Service handles setting correct defaults (QUEUED, FULL, ID generation)
  }
}